### PR TITLE
Implement webpack 4 support

### PR DESCRIPTION
**Pull request by toxic-johann**
A pull request to fix [https://github.com/GoogleChromeLabs/preload-webpack-plugin/issues/52](https://github.com/GoogleChromeLabs/preload-webpack-plugin/issues/52)

It mainly change this two part.

* Update compiler/compliation.plugin to the new Tapable api (using hook and hookAsync respectively)
* Change chunks.parents to use ChunkGroup.getParents()

But this change do not compatible webpack < 4. If we need that, we need two whole different logic, I think it's better to provide two version.

It pass 16/17 of the unit test, one failed is because `extract-text-webpack-plugin` do not support webpack4 well.

### CODE REVIEW

**Comment by jfsiii** [comment]
@toxic-johann Thanks for submitting this. 

I'm not sure which issues you were seeing with `extract-text-webpack-plugin`, but they recently released a version which fixed two issues related to webpack 4. Take a look at https://github.com/webpack-contrib/extract-text-webpack-plugin/releases/tag/v4.0.0-beta.0 and see if that helps.

**Comment by toxic-johann** [comment]
@jfsiii Thanks for the advice. The problem is fixed in the latest version.

**Comment by toxic-johann** [comment]
@jeffposnick Do you have time to review my pr? Or should i implement this based on v3?

**Comment by miduga** [comment]
👏

File: package.json [hunk 1]
```diff
@@ -30,21 +30,23 @@
...
-    "jasmine": "^2.5.3",
-    "webpack": "^3.4.0"
+    "babel-eslint": "^8.2.2",
+    "css-loader": "^0.28.10",
+    "eslint": "^4.18.2",
+    "eslint-config-google": "^0.9.1",
+    "extract-text-webpack-plugin": "^4.0.0-beta.0",
+    "file-loader": "^1.1.11",
+    "html-webpack-plugin": "^3.0.4",
+    "jasmin": "0.0.2",
```
Comment (whazap) [inline package.json]: that looks like it got added by mistake?

Comment (toxic-johann) [reply package.json]: Yes. That's a mistake.

**Comment by jeffposnick** [comment]
Thanks a ton for working on this, @toxic-johann!

I've also started work on the webpack v4 story, but I'd like to maintain legacy webpack support in the same version.

I've also been refactoring the codebase to break things out into smaller modules.

I'd love to incorporate some of the changes you made in this PR into the v3 codebase—in particular, the changes for `doesChunkBelongToHTML` was something that I was trying to figure out for myself. But, at the same time, keeping things backwards-compatible with the legacy webpack interface was my goal.

You can see where things stand with, e.g., https://github.com/GoogleChromeLabs/preload-webpack-plugin/blob/v3/src/lib/does-chunk-belong-to-html.js

So... you could either file a PR against the v3 branch with a subset of these changes, or I'm happy to extract the relevant bits around chunk groups and incorporate it as part of the work that I've been doing over the past few days (giving you credit in the code).

**Comment by toxic-johann** [comment]
Thanks, @jeffposnick . Both ways are ok. I will try to file a pr against v3 this weekend, if i have time.

**Comment by toxic-johann** [comment]
@jeffposnick Now the master branch can support both webpack v3 and webpack v4.

**Comment by jeffposnick** [comment]
Thanks a *lot* for working through this, @toxic-johann! My current plan is to merge this into `master` and then integrate the techniques into the current `v3` refactoring. I can cut another v2.x minor release in the meantime, for folks who don't want to wait on webpack 4 support.

**Comment by toxic-johann** [comment]
I agree. And I will try to make some help on `v3` branch too.

**Comment by jeffposnick** [comment]
(FYI, @toxic-johann, I've ported what I *think* is equivalent logic into the latest commits in the v3 branch. So don't feel obligated to spend time on it at the moment.)

**Comment by toxic-johann** [comment]
That's great!!

**Pull request merged**